hooks/hiero_customize_export_ui.py

```diff
@@ -0,0 +1,286 @@
+
+        Example Implementation:
+
+        .. code-block:: python
+
+            layout = widget.layout()
+            for label, prop in properties.iteritems():
+                layout.addRow(label, prop)
+
+        :param widget: The Qt widget that was created by the associated create
+            widget hook method.
+        :param OrderedDict properties: A dict containing property widget
+            objects, keyed by label, that were constructed from the data
+            built by the associated get properties hook method.
+        """
+        return
+
+    def create_transcode_exporter_widget(self, parent):
+        """
+        Builds and returns a custom widget to be embedded in the parent exporter.
+        If a custom widget is returned by this method, it will be added to the
+        parent exporter's layout.
+
+        .. note:: See the :meth:`create_shot_processor_widget` method for
+            more detailed documentation.
+
+        :param parent_widget: The parent widget.
+
+        :returns: A custom widget.
+        """
+        return None
+
+    def get_transcode_exporter_ui_properties(self):
+        """
+        Gets a list of property dictionaries describing the custom properties
+        required by the custom widget. This method will only be run if the
+        associated create widget hook method returns a widget. The dictionaries
+        will be turned into property widgets by the app before being passed to
+        the associated set properties hook method. The order that the dictionaries
+        are returned by this method is maintained when they are passed to the
+        associated set hook method.
+
+        .. note:: See the :meth:`get_shot_processor_ui_properties` method for
+            more detailed documentation.
+
+        :returns: A list of dictionaries.
+        :rtype: list
+        """
+        return []
+
+    def set_transcode_exporter_ui_properties(self, widget, properties):
+        """
+        Sets any custom properties described by get_transcode_exporter_ui_properties
+        on the custom widget returned by create_transcode_exporter_widget. This method
+        will only be called if the create method is implemented to return a custom
+        widget. The order of the properties within the dictionary passed in is the
+        same as the order they're returned in the get properties hook method.
+
+        .. note:: See the :meth:`set_shot_processor_ui_properties` method for
+            for an example implementation.
+
+        :param widget: The Qt widget that was created by the associated create
+            widget hook method.
+        :param OrderedDict properties: A dict containing property widget
+            objects, keyed by label, that were constructed from the data
+            built by the associated get properties hook method.
+        """
+        return
+
+    def create_audio_exporter_widget(self, parent):
```

Comment:
if these methods all follow the same pattern, could we have a single set of three methods where we also pass a 'type' parameter? I guess this depends on the patterns we envision people adopting - are we thinking users may override these hooks piecemeal, e.g. say just the experter widget? Or all combined? If the latter, then the type approach might work well and be a lot simpler? If the former, it makes sense to keep them separated i guess?

Reply:
I was thinking about a `type` argument as well, but I think you'd just end up with a big conditional that called out to other methods in the hook.

One idea that could simplify things a bunch would be to define a base hook that only has 3 methods (create widget and get/set properties). This could be used for any of the types, but you associate the type/hook at the config level. So if you need a custom audio and transcode hooks, you'd configure something like this:
```yaml
custom_export_ui_plugins:
  - type: audio
    hook: "{config}/hooks/tk-hiero-export/custom_audio.py"
  - type: transcode
    hook: "{config}/hooks/tk-hiero-export/custom_transcode.py"
```
Now you can cut down on the boilerplate and just define the API that's used for all the various types. Would that work?